relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,587 @@
+            let envelope = match envelope {
+                Ok(envelope) => envelope,
+                Err(err) => {
+                    relay_log::error!(
+                        error = &err as &dyn Error,
+                        "failed to unspool the envelopes from the disk",
+                    );
+                    db_error = Some(err);
+
+                    continue;
+                }
+            };
+
+            match self.extract_envelope(envelope) {
+                Ok(envelope) => {
+                    ordered_envelopes.push(OrderedEnvelope(envelope));
+                }
+                Err(err) => {
+                    relay_log::error!(
+                        error = &err as &dyn Error,
+                        "failed to extract the envelope unspooled from disk",
+                    )
+                }
+            }
+        }
+        // If there was a database error and no envelopes have been returned, we assume that we are
+        // in a critical state, so we return an error.
+        if let Some(db_error) = db_error {
+            if ordered_envelopes.is_empty() {
+                return Err(SQLiteEnvelopeStackError::DatabaseError(db_error));
+            }
+        };
+
+        // We sort envelopes by `received_at`.
+        ordered_envelopes.sort();
+
+        // We push in the back of the buffer, since we still want to give priority to
+        // incoming envelopes that have a more recent timestamp.
+        self.buffer
+            .append(&mut ordered_envelopes.into_iter().map(|o| o.0).collect());
+
+        Ok(())
+    }
+
+    /// Deserializes an [`Envelope`] from a database row.
+    fn extract_envelope(&self, row: SqliteRow) -> Result<Box<Envelope>, SQLiteEnvelopeStackError> {
+        let envelope_row: Vec<u8> = row
+            .try_get("envelope")
+            .map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+        let envelope_bytes = bytes::Bytes::from(envelope_row);
+        let mut envelope =
+            Envelope::parse_bytes(envelope_bytes).map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+
+        let received_at: i64 = row
+            .try_get("received_at")
+            .map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+        let start_time = StartTime::from_timestamp_millis(received_at as u64);
+
+        envelope.set_start_time(start_time.into_inner());
+
+        Ok(envelope)
+    }
+}
+
+impl EnvelopeStack for SQLiteEnvelopeStack {
+    type Error = SQLiteEnvelopeStackError;
+
+    async fn push(&mut self, envelope: Box<Envelope>) -> Result<(), Self::Error> {
+        if self.above_spool_threshold() {
+            // TODO: investigate how to do spooling/unspooling on a background thread.
+            self.spool_to_disk().await?;
+        }
+
+        self.buffer.push_front(envelope);
+
+        Ok(())
+    }
+
+    async fn peek(&mut self) -> Result<&Box<Envelope>, Self::Error> {
+        if self.below_unspool_threshold() {
```

Comment:
Under normal operations (when the consumers of the envelope buffer can keep up with the producers and the project state is loaded), there won't be any data on disk. So I think we should add a `read_from_disk` boolean to this struct that we set to `false` when there's nothing left on disk, such that subsequent peeks and pops don't need to query the db.

Reply:
Good idea!

Reply:
Very good suggestion, done!